lib/service/connect.go

```diff
@@ -1044,12 +1044,34 @@ func (process *TeleportProcess) rotate(conn *Connector, localState auth.StateV2,
 // falls back to trying to connect to the Auth Server through the proxy.
 // The proxy address might be configured in process environment as apidefaults.TunnelPublicAddrEnvar
 // in which case, no attempt at discovering the reverse tunnel address is made.
-func (process *TeleportProcess) newClient(authServers []utils.NetAddr, identity *auth.Identity) (*auth.Client, error) {
+func (process *TeleportProcess) newClient(authServers []utils.NetAddr, proxyAddress utils.NetAddr, identity *auth.Identity) (*auth.Client, error) {
 	tlsConfig, err := identity.TLSConfig(process.Config.CipherSuites)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	sshClientConfig, err := identity.SSHClientConfig(process.Config.FIPS)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// try and connect to the proxy server first to avoid the extra direct dial when using the value from auth servers
+	if !proxyAddress.IsEmpty() {
+		logger := process.log.WithField("proxy-address", proxyAddress.String())
+		logger.Debug("Attempting to connect to Proxy Server through tunnel.")
+
+		tunnelClient, err := process.newClientThroughTunnel([]utils.NetAddr{proxyAddress}, tlsConfig, sshClientConfig)
+		if err != nil {
+			logger.WithError(err).Debug("Failed to connect to Proxy Server through tunnel.")
+			return nil, trace.Errorf("Failed to connect to Proxy Server through tunnel, no methods remaining.")
```

Comment:
This is going to be a little confusing because:
- the error is logged here in debug level, and probably logged again by the caller
- "no methods remaining" is a bit misleading in this mode because only 1 method is ever attempted